### ISSUE TITLE
Question about using quantiles in pi 0.5 when fine tune.

### ISSUE THREAD

**Issue by emoPointer** [Physical-Intelligence/openpi#763]
Hello，

First, thank you for your excellent work！

I am currently working on **fine-tuning the pi0.5 model** using my own custom dataset on ARX arm. During this process, I've encountered a potential issue related to the `use_quantile_norm` parameter that significantly impacts performance.

### 📝 Description of the Issue

I have observed two distinct outcomes:

1.  When I manually modify the code to **disable** `use_quantile_norm` (set it to `False`), my fine-tuning process works very well, and the model achieves good performance.
2.  However, when `use_quantile_norm` is **enabled** (which appears to be the default for pi0.5), the model's performance during fine-tuning is much worse. The performance is shown in the video below.

https://github.com/user-attachments/assets/825fdfec-f5d3-44e6-abeb-8d6805903de9

| `use_quantile_norm = False` |

https://github.com/user-attachments/assets/398202c8-49a0-4c82-87ef-229c92abc720

 | `use_quantile_norm = True` |

### Evidence in Code

I investigated the codebase and found that this behavior seems to be hard-coded. In the file `openpi/src/openpi/training/config.py`, inside the `DataConfigFactory.create_base_config` method, the parameter is set as follows:

<img width="874" height="245" alt="Image" src="https://github.com/user-attachments/assets/15cefd29-0ced-4c8e-bb84-f93811e46644" />

```python
# openpi/src/openpi/training/config.py (around line 186)

    use_quantile_norm=model_config.model_type != ModelType.PI0,
)
```

This line of code automatically sets `use_quantile_norm` to **`True`** for any model that is not `ModelType.PI0` (which includes pi0.5). This prevents users from disabling it during fine-tuning without altering the core code.

### 💡 My Hypothesis

My suspicion is that this hard-coded value, while potentially correct for pre-training, may not be suitable for fine-tuning scenarios.

**My hypothesis is that for fine-tuning, `use_quantile_norm` should ideally be disabled (`False`).**

The reason is that fine-tuning datasets are often much smaller than the large-scale pre-training datasets. Applying quantile normalization to a small data distribution might be too aggressive, **potentially clipping or distorting a significant amount of useful data** (e.g., valid actions/observations). This could explain the severe performance degradation I am observing.

### 🤔 Question and Proposed Solution

My questions for the maintainers are:

* Is this hard-coded behavior intended, even for fine-tuning?
* Would you be open to making this parameter configurable, or defaulting it to `False` specifically for fine-tuning tasks?

If the team agrees that this is a valid concern, **I would be happy to submit a Pull Request** to address it.

Thank you for your time and consideration!

**Comment by zwbx** [comment]
I find the same phenomenon when I tried to finetune on brdige.

**Comment by zwbx** [comment]
your guess is interesting. I could provide some extra information, which might be helpful.
The official code does not clip action/state when do norm, so the influence can be brought by something else.

**Comment by emoPointer** [comment]
> your guess is interesting. I could provide some extra information, which might be helpful. The official code does not clip action/state when do norm, so the influence can be brought by something else.

You are right, I read the code later and there is indeed no clip data. After successfully reproducing the openpi, I didn't investigate it further, but I can provide you with more detailed configuration information.

The second video, I used assets provided by openpi and set use_quantile_norm = True.
The first video, I used assets calculated by the official script on my own data and set use_quantile_norm = False.

Furthermore, I don't believe that more than 99% of the data and less than 1% of the data will affect training. So I think my guess is wrong. I hope you resolve the problem soon.

**Comment by zwbx** [comment]
Do you suggest that it may be caused by quantile norm. I found quantile norm work worse than mean stand norm for pi05 in my case

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: ZhiqiangZhou ***@***.***>
发送时间: Friday, November 21, 2025 7:11:42 PM
收件人: Physical-Intelligence/openpi ***@***.***>
抄送: Wenbo Zhang ***@***.***>; Comment ***@***.***>
主题: Re: [Physical-Intelligence/openpi] Question about using quantiles in pi 0.5 when fine tune. (Issue #763)

[https://avatars.githubusercontent.com/u/97204642?s=20&v=4]
CAUTION: External email. Only click on links or open attachments from trusted senders.

________________________________
emoPointer left a comment (Physical-Intelligence/openpi#763)<https://github.com/Physical-Intelligence/openpi/issues/763#issuecomment-3562550898>

your guess is interesting. I could provide some extra information, which might be helpful. The official code does not clip action/state when do norm, so the influence can be brought by something else.

You are right, I read the code later and there is indeed no clip data. After successfully reproducing the openpi, I didn't investigate it further, but I can provide you with more detailed configuration information.

The second video, I used assets provided by openpi and set use_quantile_norm = True.
The first video, I used assets calculated by the official script on my own data and set use_quantile_norm = False.

Furthermore, I don't believe that more than 99% of the data and less than 1% of the data will affect training. So I think my guess is wrong. I hope you resolve the problem soon.

**Comment by emoPointer** [comment]
I guess it was caused by quantile norm before. But you know, I did not conduct the experiment under the same conditions. Asserts file in this two experiments are different too. However, if your experiment only change use_quantile_norm = True or False, Asserts file are same, and your experimental results show that set use_quantile_norm = True yields worse results, I think it's safe to say that this parameter affects performance.

**Comment by LufanM** [comment]
I find the same problem and then find your issue.(pi05 base model )
this will cause that problem which cost my a whole day to debugger. this will cause loss explosion if  one dimension value of q1 and q99 is very close, even thought you set `use_quantile_norm=False` at `TrainConfig` not in `DataConfigFactory` .
this is my norm result:
```
      "q01": [
        -0.0031515418420545753,
        -1.2550948435306548,
        -0.0017167926294729113,
        -1.568821308875084,
        -0.317965535902977,
        -0.002100556889921427,
        -0.0035556629265658555,
        0.0,
        -0.9897259978055954,
        0.9510441113948822,
        -0.04324605379998683,
        0.11724228646159171,
        -0.963572262609005,
        -0.45919464600086213,
        -0.5775401032686234,
        0.0
      ],
      "q99": [
        0.004356419151555747,
        -1.2470787916660309,
        0.005522697869874537,
        -1.5617282655715943,
        -0.31453462558984757,
        0.007054932360816748,
        0.0009483283643610778,
        0.0,
        0.2603782742023468,
        1.4846144400835037,
        1.1416017317324878,
        1.7677347258031366,
        -0.00630860440135006,
        0.8158040584325792,
        0.7555681940078736,
        0.9998
      ]
```

and here is the loss explosion log:
```
Step 0: grad_norm=1.3508, loss=0.2155, param_norm=1803.7705                                                                                         
Step 1: grad_norm=1.4429, loss=0.2140, param_norm=1803.7705                                                                                         
Step 2: grad_norm=1.8998, loss=0.2042, param_norm=1803.7705                                                                                         
Step 3: grad_norm=2.6462, loss=0.2679, param_norm=1803.7705                                                                                         
Step 4: grad_norm=229180.2656, loss=7812572160.0000, param_norm=1803.7705                                                                           
Step 5: grad_norm=2.3515, loss=0.2644, param_norm=1803.7705                                                                                         
Step 6: grad_norm=2.2156, loss=0.2213, param_norm=1803.7705                                                                                         
Step 7: grad_norm=2.8328, loss=0.3225, param_norm=1803.7705                                                                                         
Step 8: grad_norm=1.3482, loss=0.1900, param_norm=1803.7705                                                                                         
Step 9: grad_norm=2.3588, loss=0.2317, param_norm=1803.7705                                                                                         
Step 10: grad_norm=2.0265, loss=0.2306, param_norm=1803.7705                                                                                        
Step 11: grad_norm=1.8517, loss=0.2247, param_norm=1803.7705                                                                                        
Step 12: grad_norm=2.3483, loss=0.2511, param_norm=1803.7705                                                                                        
Step 13: grad_norm=1.7192, loss=0.2243, param_norm=1803.7705                                                                                        
Step 14: grad_norm=1.5450, loss=0.2265, param_norm=1803.7705                                                                                        
Step 15: grad_norm=1.5920, loss=0.2073, param_norm=1803.7705                                                                                        
Step 16: grad_norm=1.4898, loss=0.6623, param_norm=1803.7705                                                                                        
Step 17: grad_norm=1.2327, loss=0.2431, param_norm=1803.7705                                                                                        
Step 18: grad_norm=1.6556, loss=0.6448, param_norm=1803.7705                                                                                        
Step 19: grad_norm=0.8352, loss=0.1848, param_norm=1803.7705                                                                                        
Step 20: grad_norm=1.2157, loss=0.1954, param_norm=1803.7705                                                                                        
Step 21: grad_norm=1.1268, loss=0.2051, param_norm=1803.7705                                                                                        
Step 22: grad_norm=0.9039, loss=0.2028, param_norm=1803.7705                                                                                        
Step 23: grad_norm=1.0372, loss=0.1674, param_norm=1803.7705                                                                                        
Step 24: grad_norm=255020.6406, loss=7812565504.0000, param_norm=1803.7705                                                                          
Step 25: grad_norm=0.9646, loss=0.1951, param_norm=1803.7705                                                                                        
Step 26: grad_norm=0.9756, loss=0.1784, param_norm=1803.7705                                                                                        
Step 27: grad_norm=0.7181, loss=0.1650, param_norm=1803.7705                                                                                        
Step 28: grad_norm=0.7655, loss=0.1831, param_norm=1803.7705                                                                                        
Step 29: grad_norm=0.6313, loss=0.1540, param_norm=1803.7705                                                                                        
Step 30: grad_norm=0.6199, loss=0.1561, param_norm=1803.7705                                                                                        
Step 31: grad_norm=0.6857, loss=0.1875, param_norm=1803.7705                                                                                        
Step 32: grad_norm=0.6607, loss=0.1660, param_norm=1803.7705                                                                                        
Step 33: grad_norm=0.7277, loss=0.1648, param_norm=1803.7705                                                                                        
Step 34: grad_norm=201021.3594, loss=6770894336.0000, param_norm=1803.7705                                                                          
Step 35: grad_norm=0.5971, loss=0.1828, param_norm=1803.7705                                                                                        
Step 36: grad_norm=247605.9375, loss=7812572160.0000, param_norm=1803.7705                                                                          
Step 37: grad_norm=195839.6719, loss=6770890240.0000, param_norm=1803.7705                                                                          
Step 38: grad_norm=0.4831, loss=0.1699, param_norm=1803.7705          
```
